akka-http-core/src/main/scala/akka/http/impl/util/package.scala

```diff
@@ -113,6 +105,28 @@ package util {
   import akka.stream.{ Attributes, Outlet, Inlet, FlowShape }
   import scala.concurrent.duration.FiniteDuration
 
+  private[http] final case class ErrorHandling[T](handler: Throwable ⇒ Unit) extends GraphStage[FlowShape[T, T]] {
+    private val in = Inlet[T]("ErrorHandling.in")
+    private val out = Outlet[T]("ErrorHandling.out")
+
+    override val shape = FlowShape(in, out)
+
+    override def createLogic(attr: Attributes) = new GraphStageLogic(shape) {
+      setHandler(in, new InHandler {
+        override def onPush(): Unit = push(out, grab(in))
+
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          handler(ex)
+          super.onUpstreamFailure(ex)
+        }
+      })
+
+      setHandler(out, new OutHandler {
```

Comment:
Both the `InHandler` and `OutHandler` can be implemented by the anonymous `GraphStageLogic` and then just `setHanders(in, out, this)` to save two allocations on each use (we do that optimization in other graphstages)